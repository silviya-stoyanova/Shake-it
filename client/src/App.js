import React, { Component } from 'react';
import { BrowserRouter } from 'react-router-dom'
import { ToastContainer } from 'react-toastify'
import 'react-toastify/dist/ReactToastify.css';
import './static/css/site.css'
import './static/css/scrollbar.css'

import sessionManager from './components/utilities/session-util'
import observer from './components/utilities/observer'

import MyRoutes from './components/routes/router'
import Header from './components/common/header'
import Footer from './components/common/footer'

const defaultValue = { isLogged: false }
const { Consumer: UserInfoConsumer, Provider: UserInfoProvider } = React.createContext(defaultValue)

class App extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isLogged: sessionManager.isUserLogged(),
      username: sessionManager.getUsername(),
      role: sessionManager.getUserRole()
    }

    observer.subscribe('userLogin', this.onUserLogin)
    observer.subscribe('userLogout', this.onUserLogout)
  }

  onUserLogin = () => {
    this.setState(prevState => ({ // asyncronous
      isLogged: true,
      username: sessionManager.getUsername(),
      role: sessionManager.getUserRole()
    }))
  }

  onUserLogout = () => {
    this.setState(prevState => ({
      isLogged: false,
      username: '',
      role: []
    }))
  }

  render() {
    const { isLogged, username, role } = this.state

    // console.log(this.props.navigation)

    return (
      <UserInfoProvider value={{ isLogged, username, role }}>
        <BrowserRouter>
          <ToastContainer />
          <Header />

          {
            // <div className="main-container">
          }
          <MyRoutes />
          {
            // </div>
          }

          <Footer />
        </BrowserRouter>
      </UserInfoProvider>
    )
  }

  componentDidMount() {
    sessionManager.isUserLogged()
      ? this.setState({ isLogged: true })
      : this.setState({ isLogged: false })
  }
}

export { UserInfoConsumer }
export default App