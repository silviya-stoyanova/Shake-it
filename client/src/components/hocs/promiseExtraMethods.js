import requester from '../../utilities/requests-util'
import observer from '../../utilities/observer'
import sessionManager from '../../utilities/session-util'

const promiseExtraMethods = {
    user: {
        onLoginSuccess(res) {
            sessionManager.saveSession(res.authtoken, res.username, res.role)
            observer.trigger('userLogin')
        },
        onRegisterSuccess(res, data) {
            let nextPromise = requester.login(data)
            this.handleFetchPromise(nextPromise, this.onLoginSuccess, this.onLoginFail, data) // { username: data.username, password: data.password }
        },
        onLoginFail(err) {
            this.setState(prevState => ({
                data: {
                    ...prevState.data,
                    userClass: 'correct',
                    passClass: 'error',
                    repeatPassClass: 'error'
                }
            }))
        },
        onRegisterFail(err) {
            this.setState(prevState => ({
                data: {
                    ...prevState.data,
                    userClass: 'error',
                    passClass: 'correct',
                    repeatPassClass: 'correct'
                }
            }))
        }
    },

    product: {
        onProductPromiseSuccess(err) {
            return err // do nothing
        },

        onProductPromiseFail(err) {
            if (!err.titleIsTaken) {
                this.props.history.push('/')
            }
        }
    }
}

export default promiseExtraMethods