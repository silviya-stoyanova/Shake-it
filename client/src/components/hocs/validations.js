import { toast } from 'react-toastify'

const userValidations = {
    validateUserData(formType, username, password, repeatPassword, updateState) {
        let [userClass, passClass, repeatPassClass] = ['error', 'error', 'error']
        let result = {
            username: 'invalid',
            password: 'invalid',
            repeatPassword: 'invalid'
        }

        if (username && username.trim().length >= 5) {
            userClass = 'correct'
            result.username = 'valid'
        }
        if (password && password.length >= 8) {
            passClass = 'correct'
            result.password = 'valid'
        }

        if (repeatPassword) {
            if (repeatPassword && repeatPassword.length >= 8) {
                repeatPassClass = 'correct'
                result.repeatPassword = 'valid'
            }

            if (password !== repeatPassword) {
                [passClass, repeatPassClass] = ['error', 'error']
                result.password = 'invalid'
                result.repeatPassword = 'invalid'
            }
        }

        updateState && this.setState(prevState => ({
            userInfo: {
                ...prevState.userInfo,
                userClass,
                passClass,
                repeatPassClass
            }
        }))

        return result
    },

    validateUserOnSubmit(formType, username, password, repeatPassword) {
        const isInputValid = this.validateUserData(formType, username, password, repeatPassword, false)
        let isValid = false
        let errorMsg = ''

        if (formType === 'login') {
            if (isInputValid.username === 'invalid' || isInputValid.password === 'invalid') {
                errorMsg = 'Either your username or password is invalid.'
            }
        } else if (formType === 'register') {
            if (isInputValid.password === 'invalid' || isInputValid.repeatPassword === 'invalid') {
                errorMsg = 'Both passwords must consist of at least 8 characters and they both must match!'
            }
            if (isInputValid.username === 'invalid') {
                errorMsg = 'Please enter a valid username!'
            }
        }
        if (errorMsg) {
            toast.info(errorMsg, {
                className: 'error-toast'
            })
            return isValid
        }

        isValid = true
        return isValid
    }
}

const productValidations = {
    validateProductData(formType, title, description, image, price, updateState) {
        let [titleClass, descriptionClass, priceClass] = ['error', 'error', 'error', 'error']
        let result = {
            title: 'invalid',
            description: 'invalid',
            image: 'invalid',
            price: 'invalid',
        }

        if (title && title.length > 2 && title.length <= 50) {
            titleClass = 'correct'
            result.title = 'valid'
        }
        if (description && description.length > 9 && description.length <= 250) {
            descriptionClass = 'correct'
            result.description = 'valid'
        }

        // validate only when creating a new product
        if (formType === 'create' && image) {
            result.image = 'valid'
        }
        if (price && Number(price) > 0) {
            priceClass = 'correct'
            result.price = 'valid'
        }

        updateState && this.setState(prevState => ({
            productInfo: {
                ...prevState.productInfo,
                titleClass,
                descriptionClass,
                priceClass
            }
        }))
        return result
    },

    validateProductOnSubmit(formType, title, description, image, price) {
        const updateState = false
        const isInputValid = this.validateProductData(formType, title, description, image, price, updateState)
        let isValid = false
        let errorMsg = ''

        if (isInputValid.price === 'invalid') {
            errorMsg = 'Please enter a valid price.'
        }
        // when adding a product
        if (formType === 'create' && isInputValid.image === 'invalid') {
            errorMsg = 'Please provide an image.'
        }
        if (isInputValid.description === 'invalid') {
            errorMsg = 'The description must be at least 10 symbols long.'
        }
        if (isInputValid.title === 'invalid') {
            errorMsg = 'The title length must be between 3 and 50 symbols including.'
        }
        if (errorMsg) {
            toast.info(errorMsg, {
                className: 'error-toast',
            })
            return isValid
        }

        isValid = true
        return isValid
    }
}

export { userValidations }
export { productValidations }