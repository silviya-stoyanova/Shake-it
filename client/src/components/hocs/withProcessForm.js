import React, { Component } from 'react'
import { toast } from 'react-toastify'
import requester from '../../utilities/requests-util'
import sessionManager from '../../utilities/session-util'

const withProcessForm = (Form, formType, validations, initialData, requestType, promiseExtraMethods, testServiceOnMount, testServiceOnSubmit) => {
    return class extends Component {
        constructor(props) {
            super(props)
            this.state = {
                data: initialData
            }

            this.validateData = validations ? validations.validateData.bind(this) : null
            this.validateOnSubmit = validations ? validations.validateOnSubmit.bind(this) : null

            this.promiseSuccess = promiseExtraMethods ? promiseExtraMethods.success.bind(this) : null
            this.promiseFail = promiseExtraMethods ? promiseExtraMethods.fail.bind(this) : null

            this.handleInputChange = this.handleInputChange.bind(this)
            this.handleFetchPromise = this.handleFetchPromise.bind(this)
            this.handleFormSubmit = this.handleFormSubmit.bind(this)
        }

        // static defaultProps = {
        //     service: requester.getProductInfo
        // }

        async handleInputChange({ target }) {
            console.log(target.files ? target.files[0] : null)

            await this.setState(prevState => ({
                data: {
                    ...prevState.data,
                    [target.name]: target.files ? target.files : target.value, // .trim()
                    uploadedImg: target.files ? URL.createObjectURL(target.files[0]) : prevState.data.uploadedImg
                    // uploadedImg: target.files ? target.files[0] : prevState.data.uploadedImg
                }
            }))

            const updateState = true
            const { data } = this.state

            // console.log(data.uploadedImg)

            this.validateData(formType, data, updateState)
        }

        handleFetchPromise(promise, onSuccessFunc, onFailFunc, data = '') {
            promise.then(res => {
                if (!res.ok) {
                    return Promise.reject(res)
                }
                return res.json()

            }).then(res => {
                toast.info(res.success, {
                    className: 'success-toast'
                })

                // data.redirectOnSuccess &&
                this.props.history.push('/')

                onSuccessFunc && onSuccessFunc(res, data)

            }).catch(err => {
                err.json().then(error => {
                    toast.info(error.message, {
                        className: 'error-toast'
                    })

                    onFailFunc && onFailFunc(error)
                })
            })
        }

        async handleFormSubmit(event) {
            // console.log(event.preventDefault())
            
            event.preventDefault()
            const { data } = this.state
            const jwtToken = sessionManager.getUserInfo().authtoken
            let isValid = this.validateOnSubmit(formType, data)

            if (!isValid) {
                return
            }

            if (testServiceOnSubmit) {
                testServiceOnSubmit()
                    // to see the message inside the response, uncomment the next two commented out lines
                    // .then(res => res.json())
                    .then(res => {
                        // console.log(res.success)
                        this.props.history.push('/')
                    })
                    .catch(err => {
                        return err
                    })

            } else {
                const promise = requester[requestType](data, jwtToken)
                this.handleFetchPromise(promise, this.promiseSuccess, this.promiseFail, data)
            }
        }

        render() {
            return <Form {...this.state} handleInputChange={this.handleInputChange} handleFormSubmit={this.handleFormSubmit} />
        }

        componentDidMount() {
            if (formType === 'create' || formType === 'login' || formType === 'register') {
                return
            }
            // let promise = requester.getProductInfo(this.state.productInfo._id)
            // this.handleFetchPromise(promise, null, null, { redirectOnFail: true })
            // requester.getProductInfo(this.state.data._id)

            const serviceReq = testServiceOnMount || requester.getProductInfo  // new

            const productId = this.props.match                      // new
                ? this.props.match.params.productId
                : null
            // const productId = this.props.match.params.productId

            // requester.getProductInfo(productId)                  // prev
            serviceReq && serviceReq(productId)                     // new
                .then(res => {
                    if (!res.ok) {
                        return Promise.reject(res)
                    }
                    return res.json()
                })
                .then(async res => {
                    await this.setState({
                        data: res
                    })
                })
                .catch(error => {
                    this.props.history.push('/')

                    error.json().then(err => {
                        toast.info(err.message, {
                            className: 'error-toast'
                        })
                    })
                })
        }
    }
}

export default withProcessForm