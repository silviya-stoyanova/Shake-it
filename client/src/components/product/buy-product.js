import React from 'react'

const BuyProduct = () => {
    return {




    }
}

export default BuyProduct