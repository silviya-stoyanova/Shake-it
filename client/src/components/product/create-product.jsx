import React from 'react'
import withProcessProductForm from './with-product-forms-hoc'
import useTitle from '../page-title/useTitle'

const CreateProduct = (props) => {
    useTitle('Add a new product')
    const { titleClass, descriptionClass, priceClass, uploadedImg, handleInputChange, handleFormSubmit } = props

    return (
        <div className="form" >
            <form onSubmit={handleFormSubmit} encType="multipart/form-data" >
                <div className="form-type">Add a product</div>
                <hr />

                <div className="form-fields-wrapper">
                    <label htmlFor="title">Title</label>
                    <input autoFocus onChange={handleInputChange} type="text" name="title" className={titleClass} id="title" />

                    <label htmlFor="description">Description</label>
                    <textarea type="text" onChange={handleInputChange} name="description" className={descriptionClass} id="description"></textarea>

                    <div className="update-img-container">
                        <label className="input-file-container" htmlFor="image">
                            <div className="upload-img-text-wrapper">
                                <span className="upload-img-text">Upload an image</span>
                                <img className="update-img" style={{ left: '20%', top: '5%' }} src={require('../../static/images/update-pic.png')} alt="update-product" />
                            </div>

                            <div className="img-visualiser">
                                {uploadedImg
                                    ? <img src={uploadedImg} className="add-product-img" alt="product-img" />
                                    : <img src={require('../../static/images/create-product-img.png')} className="add-product-img opacity-img" alt="product-img" />
                                }
                            </div>
                        </label>
                        <input type="file" accept="image/*" onChange={handleInputChange} name="image" id="image" />
                    </div>

                    {/* <div>
                            <label className="update-img-content" htmlFor="profilePic">
                                <img className="update-img" src="" alt="update-your-profile-picture" />
                            </label>
                            <input onChange={handleInputChange} type="file" name="profilePic" id="profilePic" />
                        </div> */}


                    <label htmlFor="price">Price</label>
                    <input type="number" onChange={handleInputChange} name="price" className={priceClass} id="price" min="0" step="0.01" />
                </div>

                <hr />
                <button className="button" type="submit">Submit</button>
            </form>
        </div>
    )
}

// export default CreateProduct
export default withProcessProductForm(CreateProduct, 'create')