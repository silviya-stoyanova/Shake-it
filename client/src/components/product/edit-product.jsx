import React from 'react'
import '../../static/css/products.css'
import withProcessProductForm from './with-product-forms-hoc'
import useTitle from '../page-title/useTitle'

const EditProduct = (props) => {
    const { title, description, image, uploadedImg, price, titleClass, descriptionClass, priceClass, handleInputChange, handleFormSubmit } = props
    useTitle(`Edit ${title}`)

    return (
        <div className="form" >

            <form onSubmit={handleFormSubmit} encType="multipart/form-data">
                <div className="form-type">Edit a product</div>
                <hr />

                {image
                    ? <div className="form-fields-wrapper">
                        <input autoFocus onChange={handleInputChange} name="title" defaultValue={title} type="text" className={titleClass} id="title" />

                        <div className="update-img-container">
                            <img src={uploadedImg ? uploadedImg : `data:image/png;base64, ${image}`} alt={title} className="product-img" />

                            <div>
                                <label className="input-file-container update-img-content" htmlFor="uploadedImg">
                                    <img src={require('../../static/images/update-pic.png')} alt="update-product" className="update-img" />
                                </label>
                                <input onChange={handleInputChange} name="image" type="file" accept="image/*" id="uploadedImg" />
                            </div>
                        </div>

                        <label htmlFor="description">Description:</label>
                        <textarea onChange={handleInputChange} name="description" value={description} type="text" className={descriptionClass} id="description"></textarea>

                        <div className="price">
                            <label htmlFor="price">Price:</label>
                            <input onChange={handleInputChange} name="price" defaultValue={price} type="number" step='0.01' className={priceClass} id="price" />
                            <span className="price-sign">$</span>
                        </div>
                    </div>
                    : <img src={require('../../static/images/loading-circle.gif')} alt={'loading-img'} className="product-img" />
                }
                <hr />
                <button className="button" type="submit">Save changes</button>
            </form>
        </div>
    )
}

export default withProcessProductForm(EditProduct, 'edit')