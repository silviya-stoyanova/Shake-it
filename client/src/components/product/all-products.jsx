import React, { Component, Fragment } from 'react'
import { Link } from 'react-router-dom'
import { toast } from 'react-toastify'
import { UserInfoConsumer } from '../../App'
import requester from '../../utilities/requests-util'
import '../../static/css/products.css'

class AllProducts extends Component {
    constructor(props) {
        super(props)
        this.state = {
            products: [],
            isFetched: false
        }

        this.renderAllProducts = this.renderAllProducts.bind(this)
    }

    userFunctionalities = (data, p) => {
        return data.isLogged
            && <Fragment>
                <Link
                    to={{
                        pathname: `/product/like/${p._id}`
                    }} className="product-actions-btn">like
                </Link>
                {
                    data.role === 'Admin'
                    && <Fragment>
                        <Link
                            to={{
                                pathname: `/product/edit/${p._id}`
                            }} className="product-actions-btn">edit
                            </Link>

                        <Link
                            to={{
                                pathname: `/product/delete/${p._id}`
                            }} className="product-actions-btn">delete
                            </Link>
                    </Fragment>
                }
                <Link
                    to={{
                        pathname: `/cart/add/${p._id}`
                    }} className="product-actions-btn add-to-cart-btn">add to cart
                </Link>
            </Fragment>
    }

    renderAllProducts(data) {
        const { products } = this.state

        return (
            <Fragment>
                {
                    products.map(p => {
                        return (
                            <div key={p._id} className="product-wrapper" >
                                <Link className="product-title" to={{
                                    pathname: `/product/details/${p._id}`
                                }}>
                                    {p.title}
                                </Link>

                                <Link to={{
                                    pathname: `/product/details/${p._id}`
                                }} ><img src={'data:image/png;base64, ' + p.image} alt={p.title} className="product-img" />
                                </Link>

                                <div className="product-actions">
                                    <span className="product-price">Price: {p.price}<span className="price-sign">$</span></span>
                                    <span className="product-likes">{p.likes ? p.likes.length : 0} â™¥</span>
                                    {this.userFunctionalities(data, p)}
                                </div>
                            </div>
                        )
                    })
                }
            </Fragment >
        )
    }

    // <h1>Our site is under constructtion. We will be ready soon..</h1>
    // <h2>xoxo</h2>
    render() {
        const { products, isFetched } = this.state

        return (
            <div className='content-wrapper' >

                <UserInfoConsumer>
                    {(data) => (

                        isFetched
                            ? products.length
                                ? this.renderAllProducts(data)
                                : (<div className="no-products-container">
                                    <div className="no-products-text-container">
                                        <h1>Our products are so desired that we could not predict we would run out of stock this soon..</h1>
                                        <h2>Sorry!</h2>
                                    </div>

                                    <div className="no-products-img"></div>
                                </div>)
                            : < div className="product-wrapper" >
                                <img src={require('../../static/images/loading-circle.gif')} alt="loading-img" className="product-img" />
                            </ div>
                    )}
                </UserInfoConsumer>
            </div>
        )
    }

    componentDidMount() {
        document.title = 'Shake it - Your shake is a click away'

        requester.getAllProducts()
            .then(res => {
                if (!res.ok) {
                    return Promise.reject(res)
                }
                return res.json()
            })
            .then(allProducts => {
                this.setState({
                    products: allProducts,
                    isFetched: true
                })
            })
            .catch(err => {
                err.json()
                    .then(res => {
                        toast.info(res.message, {
                            className: 'error-toast',
                        })
                    })
            })
    }
}

export default AllProducts