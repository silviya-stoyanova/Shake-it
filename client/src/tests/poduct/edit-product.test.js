import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import { mount } from 'enzyme'
import AuthRoute from '../../components/routes/auth-route'
import { UserInfoProvider } from '../../App'
import withProcessForm from '../../components/hocs/withProcessForm'
import { EditProduct } from '../../components/product/edit-product'
import { productValidations } from '../../components/hocs/validations'

// scenarios:
//
////1. a [guest] accesses edit-product page => and then will be redirected
////2. a [user] accesses edit-product page  => and then will be redirected
////3. The [Admin] is allowed to accesses edit-product page => and then will NOT be redirected
//
//
// The [Admin] accesses edit page of [existing]-product:
//
//   -+- and the data is fetched
//4.      -- submit button is not clicked yet
////5.         -- valid info is passed to 3 of the fields
//6.           -- invalid info is passed for [name] / [img] / [description] / [price]
//7.           -- invalid info is passed for [name img description and price]
//
//8.      -- submit button is clicked
//9.           -- valid info is passed to 3 of the fields
//10.           -- invalid info is passed for [name] / [img] / [description] / [price]
//11.           -- invalid info is passed for [name img description and price]
//
//12. -+-  and the data is not fetched yet
//      /not needed to test the case when the form is submitted now, it will not be processed and will throw an error as it does on create-product, when no data is passed/
//
//
//13. The[Admin] accesses edit page of a [non-existing]-product

const initialData = {
    title: '',
    image: '',
    description: '',
    price: '',

    titleClass: 'correct',
    descriptionClass: 'correct',
    priceClass: 'correct',

    uploadedImg: ''
}

const mockExistingProductFetch = () => {
    const sampleData = JSON.stringify(
        {
            _id: 1,
            title: 'Tova e title',
            image: 'yummy-шейк.jpg',
            description: 'This is so cool, I am getting better :)',
            price: 2.93,
            likes: ['firstUserId', 'secondUserId', 'thirdUserId', 'fourthuserId', 'fifthUserId']
        })

    return Promise.resolve(new Response(sampleData))
}

const mockNonExistingProductFetch = () => {
    const sampleReturnedData = JSON.stringify({ message: 'This product does not exist!' })
    const responseOptions = { status: 404, ok: false, statusText: '' }

    return Promise.reject(
        new Response(sampleReturnedData, responseOptions)
    )
}

const historyMock = {
    push: jest.fn()
}

const sampleEvents = {
    onTitleChange: {
        target: { name: 'title', value: "new title" }
    },
    // onImageChange: {
    //     target: { name: 'image', files: ["a-file-will-be-here.png"] }
    // },
    onDescriptionChange: {
        target: { name: 'description', value: "that's right, this is a new title" }
    },
    onPriceChange: {
        target: { name: 'price', value: '229' }
    },

    onFormSubmit: {

    }
}

describe('tests for the component EditProduct', () => {

    describe('tests for guest functionalities', () => {

        it('should redirect [guests] to the home page when they accesses edit-product page', () => {
            const defaultUserValue = { isLogged: false, username: '', role: '' }

            const wrapper = mount(
                <BrowserRouter>
                    <UserInfoProvider value={defaultUserValue}>
                        <AuthRoute path="/product/edit/:productId" component={EditProduct} role="Admin" />
                    </UserInfoProvider>
                </BrowserRouter>
            )

            expect(wrapper.debug()).toMatchSnapshot()
            expect(wrapper.find('Redirect[to="/"]').length).toEqual(1)
        })
    })

    describe('tests for user functionalities', () => {

        it('should redirect [users] to the home page when they accesses edit-product page', () => {
            const defaultUserValue = { isLogged: true, username: 'my-name-is-user', role: 'User' }

            const wrapper = mount(
                <BrowserRouter>
                    <UserInfoProvider value={defaultUserValue}>
                        <AuthRoute path="/product/edit/:productId" component={EditProduct} role="Admin" />
                    </UserInfoProvider>
                </BrowserRouter>
            )

            expect(wrapper.debug()).toMatchSnapshot()
            expect(wrapper.find('Redirect[to="/"]').length).toEqual(1)
        })
    })

    describe('tests for admin functionalities', () => {

        it('should let The [Admin] accesses edit-product page by returning <Route path="/product/edit/:productId" .... />', () => {
            const defaultUserValue = { isLogged: true, username: 'xixix', role: 'Admin' }

            const wrapper = mount(
                <BrowserRouter>
                    <UserInfoProvider value={defaultUserValue}>
                        <AuthRoute path="/product/edit/:productId" component={EditProduct} role="Admin" />
                    </UserInfoProvider>
                </BrowserRouter>
            )

            expect(wrapper.debug()).toMatchSnapshot()
            expect(wrapper.find('Route[path="/product/edit/:productId"]').length).toEqual(1)
        })



        it("should render [existing] product's edit page with it's [fetched] data", () => {
            const TestHoc = withProcessForm(EditProduct, 'edit', productValidations, initialData, null, null, mockExistingProductFetch, null)

            const wrapper = mount(
                <TestHoc />
            )

            mockExistingProductFetch()
                .then(res => res.json())
                .then(res => {
                    wrapper.setState({
                        data: res
                    })

                    wrapper.update()

                    expect(wrapper.debug()).toMatchSnapshot()

                    expect(wrapper.find('div.form-type').length).toEqual(1)
                    expect(wrapper.find('div.form-type').text()).toEqual('Edit a product')

                    expect(wrapper.find('input[id="title"]').length).toEqual(1)
                    expect(wrapper.find('input[id="title"]').prop('defaultValue')).toEqual('Tova e title')

                    expect(wrapper.find('img[src="data:image/png;base64, yummy-шейк.jpg"]').length).toEqual(1)

                    expect(wrapper.find('textarea[id="description"]').text()).toEqual('This is so cool, I am getting better :)')

                    expect(wrapper.find('input[id="price"]').length).toEqual(1)
                    expect(wrapper.find('input[id="price"]').prop('defaultValue')).toEqual(2.93)

                    expect(wrapper.find('button.button[type="submit"]').text()).toEqual('Save changes')
                })
        })

        it("should mark as valid 3 of the product's fields after input change", () => {
            const TestHoc = withProcessForm(EditProduct, 'edit', productValidations, initialData, null, null, mockExistingProductFetch, null)

            const wrapper = mount(
                <TestHoc />
            )

            mockExistingProductFetch()
                .then(res => res.json())
                .then(async res => {
                    wrapper.setState({
                        data: res
                    })
                    wrapper.update()

                    await wrapper.find('input[id="title"]').simulate('change', sampleEvents.onTitleChange)
                    expect(wrapper.find('input[id="title"]').prop('defaultValue')).toEqual(sampleEvents.onTitleChange.target.value)
                    wrapper.update()
                    expect(wrapper.find('input[id="title"]').hasClass('correct')).toEqual(true)

                    await wrapper.find('textarea[id="description"]').simulate('change', sampleEvents.onDescriptionChange)
                    expect(wrapper.find('textarea[id="description"]').text()).toEqual(sampleEvents.onDescriptionChange.target.value)
                    wrapper.update()
                    expect(wrapper.find('textarea[id="description"]').hasClass('correct')).toEqual(true)

                    await wrapper.find('input[id="price"]').simulate('change', sampleEvents.onPriceChange)
                    expect(wrapper.find('input[id="price"]').prop('defaultValue')).toEqual(sampleEvents.onPriceChange.target.value)
                    wrapper.update()
                    expect(wrapper.find('input[id="price"]').hasClass('correct')).toEqual(true)
                })
        })



    })
})