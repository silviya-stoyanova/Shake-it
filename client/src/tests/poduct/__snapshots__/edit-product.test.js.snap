// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for the component EditProduct tests for admin functionalities should let The [Admin] accesses edit-product page by returning <Route path="/product/edit/:productId" .... /> 1`] = `
"<BrowserRouter>
  <Router history={{...}}>
    <AuthRoute path=\\"/product/edit/:productId\\" component={[Function: EditProduct]} role=\\"Admin\\">
      <Route path=\\"/product/edit/:productId\\" component={[Function: EditProduct]} role=\\"Admin\\" />
    </AuthRoute>
  </Router>
</BrowserRouter>"
`;

exports[`tests for the component EditProduct tests for admin functionalities should render [existing] product's edit page with it's [fetched] data 1`] = `
"<Component>
  <EditProduct data={{...}} handleInputChange={[Function: bound handleInputChange]} handleFormSubmit={[Function: bound handleFormSubmit]}>
    <div className=\\"form\\">
      <form onSubmit={[Function: bound handleFormSubmit]} encType=\\"multipart/form-data\\">
        <div className=\\"form-type\\">
          Edit a product
        </div>
        <hr />
        <div className=\\"form-fields-wrapper\\">
          <input autoFocus={true} onChange={[Function: bound handleInputChange]} name=\\"title\\" defaultValue=\\"Tova e title\\" type=\\"text\\" className={[undefined]} id=\\"title\\" />
          <div className=\\"update-img-container\\">
            <img src=\\"data:image/png;base64, yummy-шейк.jpg\\" alt=\\"Tova e title\\" className=\\"product-img\\" />
            <div>
              <label className=\\"input-file-container update-img-content\\" htmlFor=\\"uploadedImg\\">
                <img src=\\"update-pic.png\\" alt=\\"update-product\\" className=\\"update-img\\" />
              </label>
              <input onChange={[Function: bound handleInputChange]} name=\\"image\\" type=\\"file\\" accept=\\"image/*\\" id=\\"uploadedImg\\" />
            </div>
          </div>
          <label htmlFor=\\"description\\">
            Description:
          </label>
          <textarea onChange={[Function: bound handleInputChange]} name=\\"description\\" value=\\"This is so cool, I am getting better :)\\" type=\\"text\\" className={[undefined]} id=\\"description\\" />
          <div className=\\"price\\">
            <label htmlFor=\\"price\\">
              Price:
            </label>
            <input onChange={[Function: bound handleInputChange]} name=\\"price\\" defaultValue={2.93} type=\\"number\\" step=\\"0.01\\" className={[undefined]} id=\\"price\\" />
            <span className=\\"price-sign\\">
              $
            </span>
          </div>
        </div>
        <hr />
        <button className=\\"button\\" type=\\"submit\\">
          Save changes
        </button>
      </form>
    </div>
  </EditProduct>
</Component>"
`;

exports[`tests for the component EditProduct tests for guest functionalities should redirect [guests] to the home page when they accesses edit-product page 1`] = `
"<BrowserRouter>
  <Router history={{...}}>
    <AuthRoute path=\\"/product/edit/:productId\\" component={[Function: EditProduct]} role=\\"Admin\\">
      <Redirect to=\\"/\\">
        <Lifecycle onMount={[Function: onMount]} onUpdate={[Function: onUpdate]} to=\\"/\\" />
      </Redirect>
    </AuthRoute>
  </Router>
</BrowserRouter>"
`;

exports[`tests for the component EditProduct tests for user functionalities should redirect [users] to the home page when they accesses edit-product page 1`] = `
"<BrowserRouter>
  <Router history={{...}}>
    <AuthRoute path=\\"/product/edit/:productId\\" component={[Function: EditProduct]} role=\\"Admin\\">
      <Redirect to=\\"/\\">
        <Lifecycle onMount={[Function: onMount]} onUpdate={[Function: onUpdate]} to=\\"/\\" />
      </Redirect>
    </AuthRoute>
  </Router>
</BrowserRouter>"
`;
